### PR TITLE
67 吕泓江 第三次作业

### PR DESCRIPTION
**Pull request by HenryHongjiangLv**
@zyzlouis

### CODE REVIEW

File: Lesson3/assignment/yours.sol [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+如果是由owner调用：
+    function changePaymentAddress(address employeeId, address newEmployeeId) onlyOwner employeeExist(employeeId) {
+        var employee = employees[employeeId];
+        employees[newEmployeeId] = Employee(newEmployeeId, employee.salary, employee.lastPayday);
+        
+        delete employees[employeeId];
+    }
+    
+如果是员工调用：
+    function changePaymentAddress(address employeeId) onlyOwner employeeExist(msg.sender) {
```
Comment (zyzlouis) [inline Lesson3/assignment/yours.sol]: 这里的 changePaymentAddres不应该是onlyOwner，而是应该让员工自己来操作

Comment (zyzlouis) [reply Lesson3/assignment/yours.sol]: 另外请提交完整代码

File: Lesson3/assignment/yours.sol [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        var employee = employees[msg.sender];
+        employees[employeeId] = Employee(employeeId, employee.salary, employee.lastPayday);
+        
+        delete employees[msg.sender];
+    }
+
+Q3:
```
Comment (zyzlouis) [inline Lesson3/assignment/yours.sol]: 请写出继承线（过程）

**Comment by HenryHongjiangLv** [comment]
改正完成～请助教再看一下～

File: Lesson3/assignment/yours.sol [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+
+L(Z)  := [Z] + merge(L(K1), L(K2), [K1, K2])
+       = [Z] + merge([K1, A, B, O], [K2, A, C, O], [K1, K2])
+       = [Z, K1] + merge([A, B, O], [K2, A, C, O], [K2])
+       = [Z, K1, K2] + merge([A, B, O], [A, C, O])
+       = [Z, K1, K2, A] + merge([B, O], [C, O])
+       = [Z, K1, K2, A, B] + merge([O], [C, O])
+       = [Z, K1, K2, A, B, C] + merge([O], [O])
+       = [Z, K1, K2, A, B, B, O]
```
Comment (zyzlouis) [inline Lesson3/assignment/yours.sol]: C3 linearization中，有一个继承顺序你理解反了：
比如 contract K1 is A, B {...} 的意思是K1继承自B和A，同时B是离K1最近的继承关系(most-derived), 所以：

L(O) = [O]
L(A) = [A, O]
L(B) = [B, O]
L(C) = [C, O]
L(K2) = [K2] + merge(L(C), L(A), [C, A])
= [K2] + merge([C, O], [A, O], [C, A])
= [K2, C] + merge([O], [A, O], [A])
= [K2, C, A, O]
L(K1) = [K1] + merge(L(B), L(A), [B, A])
= [K1] + merge([B, O], [A, O], [B, A])
= [K1, B] + merge([O], [A, O], [A])
= [K1, B, A, O]

L(Z) = [Z] + merge(L(K2), L(K1), [K2, K1])
= [Z] + merge([K2, C, A, O], [K1, B, A, O], [K2, K1])
= [Z, K2] + merge([C, A, O], [K1, B, A, O], [K1])
= [Z, K2, C] + merge([A, O], [K1, B, A, O], [K1])
= [Z, K2, C, K1] + merge([A, O], [B, A, O])
= [Z, K2, C, K1, B] + merge([A, O], [A, O])
= [Z, K2, C, K1, B, A, O]

**Pull request merged**